schemas/com.sendgrid/processed/jsonschema/2-0-0

```diff
@@ -0,0 +1,88 @@
+{
+  "$schema": "http://iglucentral.com/schemas/com.snowplowanalytics.self-desc/schema/jsonschema/1-0-0#",
+  "description": "Schema for a SendGrid processed event",
+  "self": {
+    "vendor": "com.sendgrid",
+    "name": "processed",
+    "version": "2-0-0",
+    "format": "jsonschema"
+  },
+  "type": "object",
+  "properties": {
+    "timestamp": {
+      "description": "the UNIX timestamp of when the message was sent",
+      "type": "string",
+      "format": "date-time"
+    },
+    "email": {
+      "description": "the email address of the recipient",
+      "type": "string"
+    },
+    "newsletter": {
+      "description": "Legacy Marketing Email tool fields",
+      "type": "object",
+      "properties": {
+        "newsletter_user_list_id": {
+          "type": "string"
+        },
+        "newsletter_id": {
+          "type": "string"
+        },
+        "newsletter_send_id": {
+          "type": "string"
+        }
+      },
+      "additionalProperties": true
+    },
+    "sg_event_id": {
+      "description": "a unique ID to this event that you can use for deduplication purposes. These IDs are either 22 or 48 characters long.",
+      "type": "string",
+      "minLength": 22,
+      "maxLength": 48
+    },
+    "smtp-id": {
+      "description": "a unique ID attached to the message by the originating system",
+      "type": "string"
+    },
+    "category": {
+      "description": "Categories are custom tags that you set for the purpose of organizing your emails. Categories can be set as an array or string, and they will be returned as such when posted in your event endpoint.",
+      "type": ["array", "string"],
+      "items": {
+        "type": "string"
+      }
+    },
+    "asm_group_id": {
+      "description": "The ID of the unsubscribe group the recipient’s email address is included in. ASM IDs correspond to the ID that is returned when you create an unsubscribe group.",
+      "type": "integer",
+      "maximum": 32767,
+      "minimum": 0
+    },
+    "sg_message_id": {
+      "description": "a unique, internal SendGrid ID for the message. The first half of this is pulled from the smtp-id.",
+      "type": "string"
+    },
+    "send_at": {
+      "description": "To schedule a send request for a large batch of emails, use the send_at parameter which will send all emails at approximately the same time. send_at is a UNIX timestamp.",
+      "type": "integer",
+      "maximum": 2147483647,
+      "minimum": 0
+    },
+    "marketing_campaign_id": {
+      "description": "For emails sent through our Marketing Campaigns feature, we add Marketing Campaigns specific parameters to the Event Webhook. Both marketing_campaign_name and marketing_campaign_id are displayed as unique arguments in the event data.",
+      "type": "integer"
```

Comment:
The `marketing_campaign_*` fields are always set, they can't be `null`?

Reply:
The SendGrid [docs](https://sendgrid.com/docs/for-developers/tracking-events/event/#custom-arguments) say:

> Both `marketing_campaign_name` and  `marketing_campaign_id` are displayed as unique arguments in the event data.

I took that to mean that they are always set.

Reply:
Ok, I asked because the issue [here](https://github.com/snowplow/snowplow/issues/4002) mentions *optional* `marketing_campaign_*` fields.

What about `marketing_campaign_version` and `marketing_campaign_split_id` ?

Reply:
Ah, yes, they are optional. That is why the schema does not have any required properties. So if they are missing, their values will indeed be `null`.

Reply:
In my understanding `required` and `null` were 2 different things:
- `required`: if a field is not required, it's possible that it doesn't appear at all in the JSON.
- `null`: if `null` is not in the type (e.g. `"type": ["null", "integer"]), the value of the field can never be `null`.

Isn't it the case?

Reply:
You've got it right. If it's not required, then it can be missing from the JSON and the event will still pass validation.

If it is in the JSON, then it must match the type, which in this case is `integer`.

I meant that if it's missing, it will pass validation, but the value of the field will be `null` downstream. For example when loading into Redshift this column is nullable: https://github.com/snowplow/iglu-central/pull/905/files#diff-b05f60993433e51ea2a62f00526a47f3R20 .

Does that make sense?

Reply:
Ok so in the JSON the field will not exist unless it has a non-null value, is that right?

Reply:
Yes, that is my assumption, based on the SendGrid docs.

Reply:
Ok cool thanks!